### PR TITLE
Random Beacon v2 sortition pool locking

### PR DESCRIPTION
**Pull request by lukasz-zimnoch**
Depends on: #2671 
Depends on: https://github.com/keep-network/sortition-pools/pull/132

The sortition pool should be locked upon group creation start and unlocked either during result approval or DKG timeout.

### CODE REVIEW

**Comment by lukasz-zimnoch** [comment]
Ready for review. Keeping it draft until #2671 lands on main.

**Comment by pdyraga** [comment]
Is there a possibility we can make this PR not depending on #2671? The change looks pretty straightforward while the one from #2671 will require more work and I think there is no point in blocking this one. The merge shouldn't be that problematic I think.

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (pdyraga) [inline solidity/random-beacon/contracts/RandomBeacon.sol:630]: [RFC 19](https://github.com/keep-network/keep-core/blob/main/docs/rfc/rfc-19-random-beacon-v2.adoc) says:

>  Group creation start transaction should be embedded into relay request transaction and should lock a sortition pool. From this moment, no operator can enter or leave the pool. Once a new relay entry appears on the chain, all off-chain clients should perform group selection optimistically, using the new entry value and a view sortition pool function call. 

So the flow should be:
- Relay entry is requested. If `relay.requestCount % groupCreationFrequency == 0` and if `dkg.currentState() == DKG.State.IDLE`, we begin DKG by locking the pool.
- When relay entry is submitted, we request the creation of a new group with that relay entry. 
- DKG timeout starts counting from the moment relay entry is published.

Seed for group creation must not be known at the moment pool is locked to protect against the manipulation of sortition pool to get favorable group composition. With the current code, seed is not published to the chain but it is know - it sits in the mempool.

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (pdyraga) [inline solidity/random-beacon/contracts/RandomBeacon.sol:633]: I would not consider it as any kind of a failure, it's a normal situation.

**Review by pdyraga — CHANGES_REQUESTED** [review]

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol:633]: Yeah, this event is just in case and can make debugging easier. Do you think we should get rid of it?

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol:630]: Good point! I misunderstood the RFC. Will change.

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (pdyraga) [reply solidity/random-beacon/contracts/RandomBeacon.sol:633]: Yeah, I'd remove it to not confuse anyone. It's also some additional cost - really small but why pay it given this situation is normal. 🤷🏻

**Comment by lukasz-zimnoch** [comment]
> Is there a possibility we can make this PR not depending on #2671? The change looks pretty straightforward while the one from #2671 will require more work and I think there is no point in blocking this one. The merge shouldn't be that problematic I think.

#2671 introduce real SP in relay tests so it was easier to stack this PR on top than implementing tests using the stub and then change to real pool.

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol]: https://github.com/keep-network/keep-core/pull/2691/commits/eb221288d15eb715257eac977f68bb68efc20f5e

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol]: I have refactored the solution see: https://github.com/keep-network/keep-core/pull/2691/commits/f0a27d9d5da5eeca5344ca02271cbcbab02280d2. I'm still working on tests though. I think embedding pool lock/unlock logic within DKG state machine is the safest option. Keeping it separate seems error-prone because we can lock the pool forever or unlock it during group creation.

Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol]: Tests are ready as well: https://github.com/keep-network/keep-core/pull/2691/commits/efbe99ef5b55ea48841c74e94dc52202a7cbccc8

**Comment by lukasz-zimnoch** [comment]
@pdyraga Ready for another shot.

File: solidity/random-beacon/contracts/libraries/DKG.sol
Comment (pdyraga) [inline solidity/random-beacon/contracts/libraries/DKG.sol:75]: Maybe `AWAITING_SEED`? This is what is actually happening: DKG locked the pool and is awaiting seed from the random beacon when it's in this state.

File: solidity/random-beacon/contracts/libraries/DKG.sol
Comment (pdyraga) [inline solidity/random-beacon/contracts/libraries/DKG.sol:172]: Maybe `lockState()` or `lockSortitionPool()`? I think this would be easier to follow from the perspective of `RandomBeacon` contract. That contract also uses the pool (join / update operator), so being explicit that it's about to be locked makes sense.

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (pdyraga) [inline solidity/random-beacon/contracts/RandomBeacon.sol:460]: Can't we call `dkg.warmUp()` (or `dkg.lockState()` - see other comment) directly? We are not really creating anything here :shrug:

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (pdyraga) [inline solidity/random-beacon/contracts/RandomBeacon.sol:586]: `trigger group creation and group creation`? 😆 

```suggestion
        // If the current request should trigger group creation we need to lock
        // DKG state (lock sortition pool) to prevent operators from changing its
        // state before relay entry is known. That entry will be used as a group
        // selection seed.
```

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (pdyraga) [inline solidity/random-beacon/contracts/RandomBeacon.sol:632]: `DKG.State.AWAITING_SEED` would look really pretty here. 😉

File: solidity/random-beacon/contracts/libraries/DKG.sol
Comment (pdyraga) [inline solidity/random-beacon/contracts/libraries/DKG.sol:471]: Maybe `notifySeedTimedOut` and put this function next to `notifyTimeout`?

**Review by pdyraga — COMMENTED** [review]
First-round of comments. Still reviewing, more to come.

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol:460]: We need to call it directly in tests via `publicCreateGroupWarmUp`. Calling `dkg.warmUp` directly in the stub causes a compilation error.

File: solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts
Comment (pdyraga) [inline solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts:1178]: `tx` is not used and I believe we can also get rid of `let tx: ContractTransaction`

Comment (pdyraga) [reply solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts:1178]: Or... was the intention to add a simple `it("should work` to this describe given that we changed `it("should succeed"` into `beforeEach`? I think adding check about this event is a good move. We could check there some really basic stuff, e.g. if `DkgResultApproved` and `GroupActivated` were emitted.

File: solidity/random-beacon/contracts/libraries/DKG.sol
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/DKG.sol:172]: Yeah, good idea. I had no idea for a good nomenclature here. Thanks for the inspiration! Will change as you proposed.

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (pdyraga) [reply solidity/random-beacon/contracts/RandomBeacon.sol:460]: Let's then name it the same as `dkg.` function named, document in `@dev` we need it for tests and move to the bottom of `internal` functions.

And one day, let's get rid of the stub exposing `publicCreateGroupWarmUp` completely... 🙏🏻 🤞🏻

File: solidity/random-beacon/contracts/test/SortitionPoolStub.sol
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/test/SortitionPoolStub.sol:9]: Yep I'm aware but this should be done at once to avoid build failures 🤷‍♂️ I posted a TODO (in https://github.com/keep-network/keep-core/pull/2707) informing this stub is deprecated and real SP should be used.

File: solidity/random-beacon/test/RandomBeacon.Relay.test.ts
Comment (pdyraga) [inline solidity/random-beacon/test/RandomBeacon.Relay.test.ts:708]: To make it simpler and do not repeat checks that are not related to the warmup (awaiting seed) state, I'd do:

```
when no active groups exist after timeout is reported:
  ... put all the checks here
when no active groups exist after timeout is reported and DKG is awaiting seed:
  ... just assertions about: a) changing DKG state; b) unlocking pool
```

Comment (pdyraga) [reply solidity/random-beacon/test/RandomBeacon.Relay.test.ts:708]: And I believe assertions a) b) can be now implemented and do not require any uncomment TODO.

File: solidity/random-beacon/contracts/test/SortitionPoolStub.sol
Comment (pdyraga) [reply solidity/random-beacon/contracts/test/SortitionPoolStub.sol:9]: Yeah, it's not something we can do in this PR. I'll just keep complaining in every single PR until it's done so that we can all feel the pain lol.

File: solidity/random-beacon/contracts/libraries/DKG.sol
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/DKG.sol]: See https://github.com/keep-network/keep-core/pull/2691/commits/a29a5ad1f24125d53b37cef20129d6c5977764be

File: solidity/random-beacon/contracts/libraries/DKG.sol
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/DKG.sol]: See https://github.com/keep-network/keep-core/pull/2691/commits/a29a5ad1f24125d53b37cef20129d6c5977764be

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol]: Done as part of https://github.com/keep-network/keep-core/pull/2691/commits/a29a5ad1f24125d53b37cef20129d6c5977764be

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol]: Done as part of https://github.com/keep-network/keep-core/pull/2691/commits/a29a5ad1f24125d53b37cef20129d6c5977764be

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol]: See https://github.com/keep-network/keep-core/pull/2691/commits/a29a5ad1f24125d53b37cef20129d6c5977764be

File: solidity/random-beacon/contracts/libraries/DKG.sol
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/DKG.sol]: See https://github.com/keep-network/keep-core/pull/2691/commits/a29a5ad1f24125d53b37cef20129d6c5977764be

File: solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts
Comment (lukasz-zimnoch) [reply solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts]: Regarding the latter, it was not an intention. I just refactored that case to use `beforeEach` and be easily extensible with new `it` assertions. The basic stuff you mentioned is already checked in earlier cases above so checking them again seems redundant. Regarding usused `tx`, see: https://github.com/keep-network/keep-core/pull/2691/commits/56ead30afdaad1d6fce5f6a6d54e61ced13cbddb

File: solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts
Comment (lukasz-zimnoch) [reply solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts:1064]: See https://github.com/keep-network/keep-core/commit/56ead30afdaad1d6fce5f6a6d54e61ced13cbddb

Comment (pdyraga) [reply solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts:1064]: One more: I think we should also assert it for `describe("challengeDkgResult"`

File: solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts
Comment (pdyraga) [reply solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts]: That "should suceed" we had before was a kind of a shortcut: we don't want to repeat all previous assertions but we also want to have an explicit assertion this transaction passes at `startBlock + dkgTimeout - 1` and is supposed to do everything else as in the case tested in the earlier context.

This is also tested here because `beforeEach` would fail but... I think we miss an explicit "should work" assertion that at least suggests we expect exactly the same behavior as before. I do agree repeating the same tests does not make sense but I'd add that "should work" with at least one or two events checked so that we are not only focused on unlocking the pool but also saying the result needs to be accepted at that time, just like in any other case.

File: solidity/random-beacon/test/RandomBeacon.Relay.test.ts
Comment (lukasz-zimnoch) [reply solidity/random-beacon/test/RandomBeacon.Relay.test.ts]: See https://github.com/keep-network/keep-core/pull/2691/commits/1797b660235df348f267aabb9e9e929e07a88e15. However, assertions must be commented until we have group termination implemented. If the group is not terminated we enter in the `groups.numberOfActiveGroups() > 0` branch in `reportRelayEntryTimeout`. Once we have termination we can enter in the `else` and test the behavior of `notifySeedTimedOut`

File: solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts
Comment (lukasz-zimnoch) [reply solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts:1189]: Done as part of https://github.com/keep-network/keep-core/pull/2691/commits/2a790b469e3d5dd9e857523f9be5b6f67ac3e9bc

File: solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts
Comment (lukasz-zimnoch) [reply solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts:1064]: Done as part of https://github.com/keep-network/keep-core/pull/2691/commits/2a790b469e3d5dd9e857523f9be5b6f67ac3e9bc

File: solidity/random-beacon/contracts/libraries/DKG.sol
Comment (pdyraga) [inline solidity/random-beacon/contracts/libraries/DKG.sol:377]: `emit DkgSeedTimedOut`?

File: solidity/random-beacon/test/RandomBeacon.Relay.test.ts
Comment (pdyraga) [inline solidity/random-beacon/test/RandomBeacon.Relay.test.ts:667]: Why do we need to call these here? `publicCreateGroup` seems not to be needed and we are calling `publicDkgLockState` in the context for "and DKG is awaiting seed" at line 700.

Comment (lukasz-zimnoch) [reply solidity/random-beacon/test/RandomBeacon.Relay.test.ts]: Done in https://github.com/keep-network/keep-core/pull/2691/commits/166a6c8d38f900aa6d0e2df4c72243d870c3e25a

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol]: Done in https://github.com/keep-network/keep-core/pull/2691/commits/166a6c8d38f900aa6d0e2df4c72243d870c3e25a

File: solidity/random-beacon/contracts/libraries/DKG.sol
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/DKG.sol:181]: See https://github.com/keep-network/keep-core/pull/2691/commits/a425d466399a238a7897507eb4e00ba8061009f5

File: solidity/random-beacon/contracts/libraries/DKG.sol
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/DKG.sol:385]: See https://github.com/keep-network/keep-core/pull/2691/commits/a425d466399a238a7897507eb4e00ba8061009f5

**Comment by lukasz-zimnoch** [comment]
@pdyraga Everything addressed.

**Review by pdyraga — APPROVED** [review]

**Pull request merged**
